libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp

```diff
@@ -12,6 +15,52 @@ AP_InertialSensor_Backend::AP_InertialSensor_Backend(AP_InertialSensor &imu) :
     _sem = hal.util->new_semaphore();
 }
 
+/*
+  notify of a FIFO reset so we don't use bad data to update observed sensor rate
+ */
+void AP_InertialSensor_Backend::notify_accel_fifo_reset(uint8_t instance)
+{
+    _imu._sample_accel_count[instance] = 0;
+    _imu._sample_accel_start_us[instance] = 0;    
+}
+
+/*
+  notify of a FIFO reset so we don't use bad data to update observed sensor rate
+ */
+void AP_InertialSensor_Backend::notify_gyro_fifo_reset(uint8_t instance)
+{
+    _imu._sample_gyro_count[instance] = 0;
+    _imu._sample_gyro_start_us[instance] = 0;
+}
+
+/*
+  update the sensor rate for FIFO sensors
+
+  FIFO sensors produce samples at a fixed rate, but the clock in the
+  sensor may vary slightly from the system clock. This slowly adjusts
+  the rate to the observed rate
+*/
+void AP_InertialSensor_Backend::_update_sensor_rate(uint16_t &count, uint32_t &start_us, float &rate_hz)
+{
+    uint32_t now = AP_HAL::micros();
+    if (start_us == 0) {
+        count = 0;
+        start_us = now;
+    } else {
+        count++;
+        if (now - start_us > 1000000UL) {
+            float observed_rate_hz = count * 1.0e6 / (now - start_us);
+#if SENSOR_RATE_DEBUG
+            printf("RATE: %.1f should be %.1f\n", observed_rate_hz, rate_hz);
+#endif
+            observed_rate_hz = constrain_float(observed_rate_hz, rate_hz*0.95, rate_hz*1.05);
+            rate_hz = 0.90 * rate_hz + 0.1 * observed_rate_hz;
```

Comment:
The reason I set it so fast is that I'm seeing that the actual rates for some IMUs (eg. MPU9250 with a compass enabled when doing fast sampling) is quite a long way off from the expected rate. I'm seeing 3600Hz for an accelerometer on a MPU9250 with fast sampling with the built-in mag enabled.
I want the value to converge quickly enough that it is good before takeoff.
Note that I'm now logging the resulting observed rates.
Here are the obseved rates on an AUAV 2.1, which has a ICM20608 and an MPU9250
http://uav.tridgell.net/Paul/AUAV-sensor-rates.png
Here are the observed rates on a pixhawk1:
http://uav.tridgell.net/Paul/pixhawk1-sensor-rates.png